docs/rfc/rfc-009-consensus-parameter-upgrades.md

```diff
@@ -0,0 +1,110 @@
+did not yet have the new consensus parameters, the parameter hash it produced will
+not reference the new parameters. Any nodes joining the network with the newer
+version of Tendermint will have the new consensus parameters. Tendermint will need
+to handle this case so that new versions of Tendermint with new consensus parameters
+can still validate old blocks correctly without having to do anything overly complex
+or hacky. 
+
+### Allowing Developer Defined Values and the `EndBlock` Problem
+
+When new consensus parameters are added, application developers may wish to set
+values for them so that the developer-defined values may be used as soon as the 
+software upgrades. We do not currently have a clean mechanism for handling this.
+
+Consensus parameter updates are communicated from the application to Tendermint
+within `EndBlock` of some height `H` and take effect at the next height, `H+1`.
+This means that for updates that add a consensus parameter, there is a single
+height where the new parameters cannot take effect. The parameters did not exist
+in the version of the software that emitted the `EndBlock` response for height `H-1`,
+so they cannot take effect at height `H`. The first height that the updated params
+can take effect is height `H+1`. As of now, height `H` must run with the defaults.
+
+## Discussion
+
+### Hash Compatibility
+
+This section discusses possible solutions to the problem of maintaining backwards-compatibility 
+of hashed parameters while adding new parameters.
+
+#### Never Hash Defaults
+
+One solution to the problem of backwards-compatibility is to never include parameters
+in the hash if the are using the default value. This means that blocks produced
+before the parameters existed will have implicitly been created with the defaults.
+This works because any software with newer versions of Tendermint must be using the
+defaults for new parameters when validating old blocks since the defaults can not
+have been updated until a height at which the parameters existed.
+
+#### Only Update HashedParams on Hash-Breaking Releases
+
+An alternate solution to never hashing defaults is to not update the hashed
+parameters on non-hash-breaking releases. This means that when new consensus 
+parameters are added to Tendermint, there may be a release that makes use of the
+parameters but does not verify that they are the same across all validators by
+referencing them in the hash. This seems reasonably safe given the fact that
+only a very far subset of the consensus parameters are currently verified at all.
+
+### Developer Defined Values
+
+This section discusses possible solutions to the problem of allowing application
+developers to define values for the new parameters during the upgrade that adds
+the parameters.
+
+#### Using `InitChain` for New Values
+
+One solution to the problem of allowing application developers to define values
```

Comment:
This is a good point.

Do you think we could shim something in alongside validator set updates? Say, for example, an extra field for "consensus updates"? The down side is this would delay application by an extra round, but that's not necessarily the end of the world given we'd expect these to change fairly rarely (? I think).

Reply:
Ah, perhaps it wasn't clear, there _is_ mechanism for this alongside validator updates in `EndBlock`